sklearn/base.py

```diff
@@ -190,14 +175,17 @@ def get_params(self, deep=True):
         params : mapping of string to any
             Parameter names mapped to their values.
         """
-        out = dict()
-        for key in self._get_param_names():
-            value = getattr(self, key, None)
-            if deep and hasattr(value, 'get_params'):
-                deep_items = value.get_params().items()
-                out.update((key + '__' + k, val) for k, val in deep_items)
-            out[key] = value
-        return out
+        try:
+            param_names = get_param_names_from_constructor(self.__class__)
+        except SignatureError as e:
+            raise SignatureError("scikit-learn estimators should always "
```

Comment:
This is still more or less duplicated with kernels. Should we share this?

Reply:
> This is still more or less duplicated with kernels. Should we share this?

I left that duplication to keep the error message of each class (`scikit-learn estimators...`, `scikit-learn kernels...`). I thought of 1) having the same `scikit-learn estimators...` message for both, or 2) passing a string argument to `get_param_names_from_constructor` to edit the message, but both felt awkward.

Though looking at it again, I noticed that `set_params` is duplicated too, so maybe `Kernel` could just inherit from `BaseEstimator` . Are there any historical reasons as to why it does not does so? By inheriting It could also benefit from the pickling and repr stuff.

Also, Kernels do not include any nested estimator-like parameters right? `get_params` is only shallow which seems enough, but `set_params` can handle nested parameters, why is this?